src/briefcase/platforms/linux/appimage.py

```diff
@@ -157,49 +140,77 @@ def support_package_url(self, support_revision):
             + self.support_package_filename(support_revision)
         )
 
-    def _pip_requires(self, requires: List[str]):
-        """Convert the requirements list to an AppImage compatible format.
-
-        If running in no-docker mode, this returns the requirements as specified
-        by the user.
+    def _install_app_requirements(
+        self,
+        app: AppConfig,
+        requires: List[str],
+        app_packages_path: Path,
+    ):
+        """Install requirements for the app with pip.
 
-        If running in Docker mode, any local file requirements are converted into
-        a reference in the /requirements mount folder.
+        This method pre-compiles any requirement defined using a local path
+        reference into an sdist tarball. This will be used when installing under
+        Docker, as local file references can't be accessed in the Docker
+        container.
 
-        :param requires: The user-specified list of app requirements
-        :returns: The final list of requirement arguments to pass to pip
+        :param app: The app configuration
+        :param requires: The list of requirements to install
+        :param app_packages_path: The full path of the app_packages folder into
+            which requirements should be installed.
         """
-        original = super()._pip_requires(requires)
-        if not self.use_docker:
-            return original
-
-        final = []
-        for requirement in original:
+        # If we're re-building requirements, purge any pre-existing local
+        # requirements.
+        local_requirements_path = self.local_requirements_path(app)
+        if local_requirements_path.exists():
+            self.tools.shutil.rmtree(local_requirements_path)
+        self.tools.os.mkdir(local_requirements_path)
+
+        # Iterate over every requirements, looking for local references
+        for requirement in requires:
             if _is_local_requirement(requirement):
-                source = Path(requirement).resolve()
-                final.append(f"/requirements/{source.name}")
-            else:
-                # Use the requirement as-specified
-                final.append(requirement)
-
-        return final
+                # Build an sdist for the local requirement
+                with self.input.wait_bar(f"Building sdist for {requirement}..."):
+                    self.tools.subprocess.check_output(
+                        [
```

Comment:
> I've swallowed the output of generating the sdist in the wait-bar. Build output is captured in the log, but isn't surfaced for the normal successful build use case.

The first time I tried this with the testbed app, I forgot to install the new requirement for `build`. The correct error message did appear on screen, albeit on the same line as the wait bar:
```
[testbed] Updating requirements...
     ━━━━━━━━━━━━━━━━━━━━ Building sdist for ../core.../home/smith/.venv/bw-38/bin/python3.8: No module named build
Building sdist for ../core...
```
But it didn't appear anywhere in the log:
```
           >>> Running Command:                                                                                                                                    subprocess.py:660
           >>>     /home/smith/.venv/bw-38/bin/python3.8 -m build --sdist --outdir '/home/smith/git/beeware/toga/testbed/linux/appimage/Toga Testbed/_local'       subprocess.py:661
           ../core                                                                                                                                                                  
           >>> Working Directory:                                                                                                                                  subprocess.py:668
           >>>     /home/smith/git/beeware/toga/testbed                                                                                                            subprocess.py:669
           >>> Return code: 1                                                                                                                                      subprocess.py:696
           Building sdist for ../core...                                                                                                                             appimage.py:173

Main thread traceback:
...
```

Later, when I discovered the problem with existing sdists or wheels, that error message appeared in the log but not on the screen. Maybe the difference is down to stdout vs stderr?

Reply:
stdout/err is definitely the cause here. @rmartin16 has gotten ahead of this and submitted #1059; but even with that change, I can improve the error reporting here.